scripts/change-log-generator-utils.js

```diff
@@ -0,0 +1,274 @@
+          util.format(MESSAGE_FORMAT, map[MESSAGE], map[PR_NUM], map[PR_NUM])
+        );
+      }
+    });
+  });
+  Object.keys(groupedMessages)
+    .sort()
+    .forEach(function(key) {
+      sortedMessages[key] = groupedMessages[key];
+    });
+  return sortedMessages;
+}
+
+function getChangeLogText(releaseBranch, groupedMessages) {
+  let changeLogText = util.format(
+    LOG_HEADER,
+    releaseBranch.toString().replace(constants.RELEASE_BRANCH_PREFIX, ''),
+    getReleaseDate()
+  );
+  let lastType = '';
+  Object.keys(groupedMessages).forEach(function(typeAndPackageName) {
+    let [type, packageName] = typeAndPackageName.split('|');
+    if (!lastType || lastType != type) {
+      changeLogText += util.format(TYPE_HEADER, type);
+      lastType = type;
+    }
+    changeLogText += util.format(SECTION_HEADER, packageName);
+    groupedMessages[typeAndPackageName].forEach(function(message) {
+      changeLogText += message;
+    });
+  });
+  return changeLogText + '\n';
+}
+
+function getFilesChanged(commitNumber) {
+  return shell
+    .exec('git show --pretty="" --name-only ' + commitNumber, {
+      silent: true
+    })
+    .stdout.trim()
+    .toString()
+    .replace(/\n/g, ',');
+}
+
+function getPackageHeaders(filesChanged) {
+  let packageHeaders = new Set();
+  filesChanged.split(',').forEach(function(filePath) {
+    const packageName = getPackageName(filePath);
+    if (packageName) {
+      packageHeaders.add(packageName);
+    }
+  });
+  return filterPackageNames(packageHeaders);
+}
+
+function writeChangeLog(textToInsert) {
+  let data = fs.readFileSync(constants.CHANGE_LOG_PATH);
+  let fd = fs.openSync(constants.CHANGE_LOG_PATH, 'w+');
+  let buffer = Buffer.from(textToInsert.toString());
+  fs.writeSync(fd, buffer, 0, buffer.length, 0);
+  fs.writeSync(fd, data, 0, data.length, buffer.length);
+  fs.closeSync(fd);
+  console.log(`\nChange log written to: ${constants.CHANGE_LOG_PATH}`);
+}
+
+function getPackageName(filePath) {
+  if (
+    filePath &&
+    !filePath.includes('/images/') &&
+    !filePath.includes('/test/')
+  ) {
+    let packageName = filePath.replace('packages/', '').split('/')[0];
+    return packageName.startsWith('salesforce') ||
+      packageName.startsWith('docs')
+      ? packageName
+      : null;
+  }
+  return null;
+}
+
+function filterPackageNames(packageHeaders) {
+  let filteredHeaders = new Set(packageHeaders);
+  if (packageHeaders.has('salesforcedx-vscode-core')) {
+    packageHeaders.forEach(function(packageName) {
+      if (packageName != 'salesforcedx-vscode-core' && packageName != 'docs') {
+        filteredHeaders.delete(packageName);
+      }
+    });
+  }
+  return filteredHeaders;
+}
+
+/**
+ * Generate the key to be used in the grouped messages map. This will help us
+ * determine whether this is an addition or fix, along with the package header
+ * that the commit should be inserted under.
+ *
+ * If we have a type that should be ignored, return an empty key.
+ */
```

Comment:
😂  jsdoc FTW. Should you go ahead and include the params? more vscode friendly

Reply:
:) I'll add them to other functions as well.